django/utils/encoding.py

```diff
@@ -196,20 +193,45 @@ def iri_to_uri(iri):
     return quote(force_bytes(iri), safe=b"/#%[]=:;$&()+,!?*@'~")
 
 
+# List of byte values that uri_to_iri is going to decode from percent encoding.
+# These are the unreserved characters from RFC 3986 and everything above 128,
+# because bytes >= 128 are part of multibyte unicode characters.
+_decode_ranges = [[45, 46, 95, 126], range(65, 91), range(97, 123), range(128, 256)]
+_hextobyte = {
+    (fmt % char).encode(): six.int2byte(char)
+    for range in _decode_ranges for char in range for fmt in ['%02x', '%02X']
+}
+
+
 def uri_to_iri(uri):
     """
     Converts a Uniform Resource Identifier(URI) into an Internationalized
     Resource Identifier(IRI).
 
-    This is the algorithm from section 3.2 of RFC 3987.
+    This is the algorithm from section 3.2 of RFC 3987, excluding step 4.
```

Comment:
Interesting! I guess we could implement it at some point, but that seems like a fair amount of work.